### ISSUE TITLE
Use CustomizeDiff to infer the provider default values

### ISSUE THREAD

**Issue by rileykarson** [hashicorp/terraform-provider-google#11339]
### Affected Resource(s)

* google_*

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

The provider default project (`project` field) is substituted in for resources whose `project` field is empty at apply time across the provider using the `getProject` method, which checks state and then the provider config for the project. The `config` object holding provider configuration was only available at apply time, historically, but that's not true with `CustomizeDiff` around nowadays. We could, presumably, implement a standardised `CustomizeDiff` for all project-level resources, and determine the final value of the `project` field (and `zone` + `region`) at plan time. Given a config like the following:

```
provider "google" {
  project = "my-project"
}

resource "google_compute_address" "my-address" {}
```

Our provider would go from the following `terraform plan` on creation:

```
...
      + project            = (known after apply)
...
```

To the following:

```
...
      + project            = "my-project"
...
```

Additionally, we could return a plan-time error when the value is not set.

Note that the current behaviour where the project value is "sticky" was originally due to not having enough information at plan time. We could preserve the existing behaviour by only rewriting the new state to the provider project when both old and new state are empty, and eliminate the behaviour by rewriting whenever new is empty. That's clearly a breaking change, however, I feel that adding the plan-time display+error likely counts as one as well.

In short, I think we could use CustomizeDiff to improve our handling of default projects (but it would be a breaking change to do either of these):
* Writing the provider project into the new state on create when the resource-level one is unset, displaying it in plan, and adding a check to return an error if it is empty
* Writing the provider project into the new state on all plans when the resource-level one is unset, overriding old values and removing the current sticky behaviour

Also, as a note: we'd need to check for unknown values in CustomizeDiff! CD does *not* protect against them by default, in contrast to other schema functions, and unknown values appear to be empty without special handling.

**Comment by wyardley** [comment]
Thanks Riley - that would be super helpful!

**Comment by wyardley** [comment]
Maybe related: #5761?

**Comment by rileykarson** [comment]
Note: the actual change (or changing just generated resources) is likely size/s or size/m, but applying to all resources makes this size/xl

**Comment by c2thorn** [comment]
closed with https://github.com/GoogleCloudPlatform/magic-modules/pull/8569

This will release with v5.0.0. Check https://github.com/hashicorp/terraform-provider-google/issues/15582 for more details on the major release.
